File: deno/src/wx/wx_pay_notice/wx_pay_notice.service.ts
import {
  error,
  log,
} from "/lib/context.ts";

import {
  getWxPayModel,
} from "/src/wx/wx_pay/wx_pay.dao.ts";

import {
  findOne as findOneWxPay,
} from "/gen/wx/wx_pay/wx_pay.dao.ts";

import {
  create as createWxPayNotice,
  updateById as updateByIdWxPayNotice,
} from "/gen/wx/wx_pay_notice/wx_pay_notice.dao.ts";

import {
  findOne as findOnePayTransactionsJsapi,
  validateOption as validateOptionTransactionsJsapi,
  updateById as updateByIdTransactionsJsapi,
} from "/gen/wx/pay_transactions_jsapi/pay_transactions_jsapi.dao.ts";

import dayjs from "dayjs";

import {
  Decimal,
} from "decimal.js";

import {
  isEmpty,
} from "/lib/util/string_util.ts";

import WxPay from "wechatpay-node-v3";

import {
  WxPayNoticeTradeType,
  WxPayNoticeTradeState,
  WxPayNoticeCurrency,
  WxPayNoticePayerCurrency,
  PayTransactionsJsapiTradeState,
} from "/gen/types.ts";

import {
  Buffer,
} from "node:buffer";

interface Ipay {
  appid: string; //  直连商户申请的公众号或移动应用appid。
  mchid: string; // 商户号
  serial_no?: string; // 证书序列号
  publicKey: string; // 公钥
  privateKey: string; // 密钥
  authType?: string; // 认证类型，目前为WECHATPAY2-SHA256-RSA2048
  userAgent?: string;
  key?: string;
}

/*
{
  "mchid": "1550006391",
  "appid": "wx28e52faff22e8837",
  "out_trade_no": "443aafc9179549ddacb75c95efdb7799",
  "transaction_id": "4200001740202302084175053914",
  "trade_type": "JSAPI",
  "trade_state": "SUCCESS",
  "trade_state_desc": "支付成功",
  "bank_type": "OTHERS",
  "attach": "",
  "success_time": "2023-02-08T18:52:29+08:00",
  "payer": {
    "openid": "o2tur4paomwplOlVeKL8z-cDkKFE"
  },
  "amount": {
    "total": 1,
    "payer_total": 1,
    "currency": "CNY",
    "payer_currency": "CNY"
  }
}
*/
interface PayNoticeBody {
  mchid: string;
  appid: string;
  out_trade_no: string;
  transaction_id: string;
  trade_type: string;
  trade_state: string;
  trade_state_desc: string;
  bank_type: string;
  attach: string;
  success_time: string;
  payer: {
    openid: string;
  };
  amount: {
    total: number;
    payer_total: number;
    currency: string;
    payer_currency: string;
  };
  device_id: string;
}

async function fetchPay_notice(
  obj: Ipay,
  params: {
    // deno-lint-ignore no-explicit-any
    body: any;
    signature: string;
    serial: string;
    nonce: string;
    timestamp: string;
  },
) {
  const wxPay = new WxPay({
    appid: obj.appid,
    mchid: obj.mchid,
    publicKey: Buffer.from(obj.publicKey),
    privateKey: Buffer.from(obj.privateKey),
    key: obj.key,
  });
  const ret = await wxPay.verifySign(params);
  if (!ret) {
    throw new Error("签名验证失败");
  }
  const pay_noticeBody = params.body;
  // 数据密文
  const ciphertext = pay_noticeBody.resource.ciphertext;
  // 附加数据
  const associated_data = pay_noticeBody.resource.associated_data;
  // 随机串
  const nonce = pay_noticeBody.resource.nonce;
  // APIv3密钥
  const key = obj.key;
  const data = wxPay.decipher_gcm<PayNoticeBody>(ciphertext, associated_data, nonce, key);
  return data;
}

/**
 * 支付通知
**/
export async function pay_notice(
  notify_url: string,
  params: {
    // deno-lint-ignore no-explicit-any
    body: any;
    signature: string;
    serial: string;
    nonce: string;
    timestamp: string;
  },
) {
  log(`pay_notice_params: ${ JSON.stringify(params) }`);
  /*
  {
    "id": "19784e48-5800-57c1-a9e7-ea5ec6cc3e2c",
    "create_time": "2023-02-08T18:50:07+08:00",
    "resource_type": "encrypt-resource",
    "event_type": "TRANSACTION.SUCCESS",
    "summary": "支付成功",
    "resource": {
      "original_type": "transaction",
      "algorithm": "AEAD_AES_256_GCM",
      "ciphertext": "goMqsaZa+OH/70QLKZFh7hl+mXkI6V9HGkHFOQw7V4f9l2zm1JdVZeiVXnUuhfo3vMlklZGztjdGd+wi8TF8afC5BFB1v2TG6g+pr6h0xUU2r7sjaG4Cv2ZfWiRrB1guA6UZH1nkrVCsO61hb8uuq0OMMuWNCxGGAWrYg5A1CHsdAeQPbiAyxE+hlkt6cRonp20fSnrsk1ofoiBAPs+I1PGG9w5WOfuy+kgvjLRwLXWv6QFyMkSijFUYCTrVmp7U8833x8NdPDfone1nsAUlHgXidGvvC3QKGhuBTycLHiC8EcwFGcktzUUSKucmXSeI/ggptKUxwmoLoHBUo49x8PKQj9oiYNiPhQbm7rHJPC5LuF/u3HaiNV7ijRgvJuJ1Ctk1KuxlxCes8YGUASuxPmrcgVEpZQXvzIV5w6wrIJ+4lcFSUtycDU0rfNC77usJT7bjpndi+7opdEbjf488CcGIRy7EJ7YL2uqlXDEnZGXuY3Kl/o4w4yspiTTfm6Qo633crhH9DQ4DRQj45gsCIYmP8VjK/HkaG0mUzLF3gqe8nxgh9CWgMLOCsDVtHpiCfspAOfFk2cQrfqC/f9a9",
      "associated_data": "transaction",
      "nonce": "p1dNx5tmwsyk"
    }
  }
  */
  log(`pay_notice_body: ${ JSON.stringify(params.body) }`);
  const wx_payModel0 = await findOneWxPay({
    notify_url,
  });
  if (!wx_payModel0 || !wx_payModel0.appid) {
    error(`回调地址: ${ notify_url } 未配置微信支付`);
    return;
  }
  const wx_payModel = await getWxPayModel(wx_payModel0);
  const appid = wx_payModel.appid;
  const mchid = wx_payModel.mchid;
  const public_key = wx_payModel.public_key;
  const private_key = wx_payModel.private_key;
  const v3_key = wx_payModel.v3_key;
  
  const result = await fetchPay_notice(
    {
      appid,
      mchid,
      publicKey: public_key,
      privateKey: private_key,
      key: v3_key,
    },
    params,
  );
  log(`pay_notice_result: ${ JSON.stringify(result) }`);
  let success_time: string | undefined = undefined;
  if (result.success_time) {
    success_time = dayjs(result.success_time).format("YYYY-MM-DD HH:mm:ss");
  }
  /*
  {
    "mchid": "1550006391",
    "appid": "wx28e52faff22e8837",
    "out_trade_no": "443aafc9179549ddacb75c95efdb7799",
    "transaction_id": "4200001740202302084175053914",
    "trade_type": "JSAPI",
    "trade_state": "SUCCESS",
    "trade_state_desc": "支付成功",
    "bank_type": "OTHERS",
    "attach": "",
    "success_time": "2023-02-08T18:52:29+08:00",
    "payer": {
      "openid": "o2tur4paomwplOlVeKL8z-cDkKFE"
    },
    "amount": {
      "total": 1,
      "payer_total": 1,
      "currency": "CNY",
      "payer_currency": "CNY"
    }
  }
  */
  // 记录订单
  const wx_pay_noticeId: WxPayNoticeId = await createWxPayNotice({
    appid: result.appid,
    mchid: result.mchid,
    openid: result.payer.openid,
    out_trade_no: result.out_trade_no,
    transaction_id: result.transaction_id,
    trade_type: result.trade_type as unknown as WxPayNoticeTradeType,
    trade_state: result.trade_state as unknown as WxPayNoticeTradeState,
    trade_state_desc: result.trade_state_desc,
    bank_type: result.bank_type,
    attach: result.attach,
    success_time,
    total: result.amount.total,
    payer_total: result.amount.payer_total,
    currency: result.amount.currency as unknown as WxPayNoticeCurrency,
    payer_currency: result.amount.payer_currency as unknown as WxPayNoticePayerCurrency,
    device_id: result.device_id,
    raw: JSON.stringify(result),
  });
  const transaction_id = result.transaction_id;
  const trade_state = result.trade_state as unknown as PayTransactionsJsapiTradeState;
  const trade_type = result.trade_type;
  const out_trade_no = result.out_trade_no;
  const trade_state_desc = result.trade_state_desc;
  const total = result.amount.total;
  
  const pay_transactions_jsapiModel = await validateOptionTransactionsJsapi(
    await findOnePayTransactionsJsapi({
      out_trade_no,
    }),
  );
  const pay_transactions_jsapi_id: PayTransactionsJsapiId = pay_transactions_jsapiModel.id;
  const tenant_id: TenantId = pay_transactions_jsapiModel.tenant_id;
  const attach2 = pay_transactions_jsapiModel.attach2;
  
  if (trade_state === "SUCCESS" && trade_type === "JSAPI") {
    await updateByIdWxPayNotice(
      wx_pay_noticeId,
      {
        tenant_id,
      },
    );
    
    await updateByIdTransactionsJsapi(
      pay_transactions_jsapi_id,
      {
        transaction_id,
        trade_state,
        trade_state_desc,
        success_time,
      },
    );
  } else {
    error(`支付回调失败: ${ JSON.stringify(result) }`);
    return;
  }
  if (isEmpty(attach2)) {
    error(`pay_notice_attach2: attach2 为空!`);
    return;
  }
  const attach2Obj = JSON.parse(attach2);
  const action = attach2Obj.action;
  
  const amt = new Decimal(total / 100);
  const integral = Math.floor(total / 100);
  // if (action === "recharge_card") {
  // }
}

File: rust/src/src/wx/wx_pay_notice/wx_pay_notice_service.rs
use color_eyre::eyre::{Result, eyre};
use tracing::{info, error};

use rust_decimal::prelude::Decimal;

use crate::common::context::{
  Options,
  get_req_id,
};

use crate::common::wx_pay::decode::{
  decode_wx_pay,
  WxPayNotify,
  WxPayResource,
};
use crate::common::wx_pay::{
  TradeType,
  TradeState,
};

// wx_pay
use crate::r#gen::wx::wx_pay::wx_pay_dao::{
  find_one_wx_pay,
  validate_option_wx_pay,
};
use crate::r#gen::wx::wx_pay::wx_pay_model::WxPaySearch;

// wx_pay_notice
use crate::r#gen::wx::wx_pay_notice::wx_pay_notice_dao::{
  create_wx_pay_notice,
  update_tenant_by_id_wx_pay_notice,
};
use crate::r#gen::wx::wx_pay_notice::wx_pay_notice_model::{
  WxPayNoticeTradeType,
  WxPayNoticeTradeState,
  WxPayNoticeCurrency,
  WxPayNoticePayerCurrency,
  WxPayNoticeInput,
};

// pay_transactions_jsapi
use crate::r#gen::wx::pay_transactions_jsapi::pay_transactions_jsapi_dao::{
  find_one_pay_transactions_jsapi,
  validate_option_pay_transactions_jsapi,
  update_by_id_pay_transactions_jsapi,
};
use crate::r#gen::wx::pay_transactions_jsapi::pay_transactions_jsapi_model::{
  PayTransactionsJsapiTradeState,
  PayTransactionsJsapiSearch,
  PayTransactionsJsapiInput,
};

static NOTIFY_URL: &str = "/api/wx_pay/wx_pay_notify";

pub async fn wx_pay_notify(
  wx_pay_notify: WxPayNotify,
  options: Option<Options>,
) -> Result<()> {
  
  info!(
    "{req_id} wx_pay_notify: {wx_pay_notify:?}",
    req_id = get_req_id(),
  );
  
  let wx_pay_model = validate_option_wx_pay(
    find_one_wx_pay(
      Some(WxPaySearch {
        notify_url: Some(NOTIFY_URL.to_string()),
        ..Default::default()
      }),
      None,
      options.clone(),
    ).await?,
  ).await?;
  
  let appid = wx_pay_model.appid;
  let mchid = wx_pay_model.mchid;
  let v3_key = wx_pay_model.v3_key;
  
  let raw = serde_json::to_string(&wx_pay_notify)?;
  
  let wx_pay_resource: WxPayResource = decode_wx_pay(
    &v3_key,
    wx_pay_notify,
  )?;
  
  info!(
    "{req_id} wx_pay_resource: {wx_pay_resource:?}",
    req_id = get_req_id(),
  );
  
  let openid = wx_pay_resource.payer.openid;
  let out_trade_no = wx_pay_resource.out_trade_no;
  let transaction_id = wx_pay_resource.transaction_id;
  let trade_type: WxPayNoticeTradeType = match wx_pay_resource.trade_type {
    TradeType::JSAPI => WxPayNoticeTradeType::Jsapi,
    TradeType::NATIVE => WxPayNoticeTradeType::Native,
    TradeType::APP => WxPayNoticeTradeType::App,
    TradeType::MICROPAY => WxPayNoticeTradeType::Micropay,
    TradeType::MWEB => WxPayNoticeTradeType::Mweb,
    TradeType::FACEPAY => WxPayNoticeTradeType::Facepay,
  };
  let trade_state: WxPayNoticeTradeState = match wx_pay_resource.trade_state {
    TradeState::SUCCESS => WxPayNoticeTradeState::Success,
    TradeState::REFUND => WxPayNoticeTradeState::Refund,
    TradeState::NOTPAY => WxPayNoticeTradeState::Notpay,
    TradeState::CLOSED => WxPayNoticeTradeState::Closed,
    TradeState::REVOKED => WxPayNoticeTradeState::Revoked,
    TradeState::USERPAYING => WxPayNoticeTradeState::Userpaying,
    TradeState::PAYERROR => WxPayNoticeTradeState::Payerror,
  };
  let trade_state_desc = wx_pay_resource.trade_state_desc;
  let bank_type = wx_pay_resource.bank_type;
  let attach = wx_pay_resource.attach.unwrap_or_default();
  //  "2018-06-08T10:34:56+08:00"
  let success_time = wx_pay_resource.success_time;
  let success_time = chrono::NaiveDateTime::parse_from_str(&success_time, "%Y-%m-%dT%H:%M:%S%z")?;
  let total = wx_pay_resource.amount.total;
  if total > u32::MAX as u64 {
    return Err(eyre!("total is too large"));
  }
  let total = total as u32;
  let payer_total = wx_pay_resource.amount.payer_total;
  if payer_total > u32::MAX as u64 {
    return Err(eyre!("payer_total is too large"));
  }
  let payer_total = payer_total as u32;
  let currency = match wx_pay_resource.amount.currency.as_str() {
    "CNY" => Some(WxPayNoticeCurrency::Cny),
    _ => None,
  };
  let payer_currency = wx_pay_resource.amount.payer_currency;
  let payer_currency: Option<WxPayNoticePayerCurrency> = match payer_currency.as_str() {
    "CNY" => Some(WxPayNoticePayerCurrency::Cny),
    _ => None,
  };
  let device_id = wx_pay_resource.scene_info
    .map(|scene_info|
      scene_info.device_id.unwrap_or_default()
    );
  
  let wx_pay_notice_id = create_wx_pay_notice(
    WxPayNoticeInput {
      appid: Some(appid),
      mchid: Some(mchid),
      openid: Some(openid),
      out_trade_no: Some(out_trade_no.clone()),
      transaction_id: Some(transaction_id.clone()),
      trade_type: Some(trade_type),
      trade_state: Some(trade_state),
      trade_state_desc: Some(trade_state_desc.clone()),
      bank_type: Some(bank_type),
      attach: Some(attach),
      success_time: Some(success_time.clone()),
      total: Some(total),
      payer_total: Some(payer_total),
      currency,
      payer_currency,
      device_id,
      raw: Some(raw),
      ..Default::default()
    },
    options.clone(),
  ).await?;
  
  let pay_transactions_jsapi_model = validate_option_pay_transactions_jsapi(
    find_one_pay_transactions_jsapi(
      Some(PayTransactionsJsapiSearch {
        out_trade_no: Some(out_trade_no),
        ..Default::default()
      }),
      None,
      options.clone(),
    ).await?,
  ).await?;
  
  let pay_transactions_jsapi_id = pay_transactions_jsapi_model.id;
  let tenant_id = pay_transactions_jsapi_model.tenant_id;
  let attach2 = pay_transactions_jsapi_model.attach2;
  
  update_tenant_by_id_wx_pay_notice(
    wx_pay_notice_id,
    tenant_id,
    options.clone(),
  ).await?;
  
  let trade_state = match trade_state {
    WxPayNoticeTradeState::Success => PayTransactionsJsapiTradeState::Success,
    WxPayNoticeTradeState::Refund => PayTransactionsJsapiTradeState::Refund,
    WxPayNoticeTradeState::Notpay => PayTransactionsJsapiTradeState::Notpay,
    WxPayNoticeTradeState::Closed => PayTransactionsJsapiTradeState::Closed,
    WxPayNoticeTradeState::Revoked => PayTransactionsJsapiTradeState::Revoked,
    WxPayNoticeTradeState::Userpaying => PayTransactionsJsapiTradeState::Userpaying,
    WxPayNoticeTradeState::Payerror => PayTransactionsJsapiTradeState::Payerror,
  };
  
  update_by_id_pay_transactions_jsapi(
    pay_transactions_jsapi_id,
    PayTransactionsJsapiInput {
      transaction_id: Some(transaction_id),
      trade_state: Some(trade_state),
      trade_state_desc: Some(trade_state_desc),
      success_time: Some(success_time),
      ..Default::default()
    },
    options.clone(),
  ).await?;
  
  if trade_state != PayTransactionsJsapiTradeState::Success {
    return Ok(());
  }
  
  if attach2.is_empty() {
    error!(
      "{req_id} pay_notice_attach2: attach2 is empty",
      req_id = get_req_id(),
    );
    return Ok(());
  }
  
  let attach2_obj: serde_json::Value = serde_json::from_str(&attach2)?;
  let action = attach2_obj["action"].as_str().unwrap_or_default();
  
  let amt = Decimal::from(payer_total) / Decimal::from(100);
  
  info!(
    "{req_id} pay_notice_attach2: action: {action} amt: {amt} {attach2:?}",
    req_id = get_req_id(),
    attach2 = attach2_obj,
  );
  
  // TODO: handle action
  
  Ok(())
}
